### PR TITLE
rollup: add `rollup/circuitcapacitychecker` package

### PR DESCRIPTION
**Pull request by 0xmountaintop**
## 1. Purpose or design rationale of this PR

...

## 2. PR title

Your PR title must follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) (as we are doing squash merge for each PR), so it must start with one of the following [types](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type):

- [ ] build: Changes that affect the build system or external dependencies (example scopes: yarn, eslint, typescript)
- [ ] ci: Changes to our CI configuration files and scripts (example scopes: vercel, github, cypress)
- [ ] docs: Documentation-only changes
- [ ] feat: A new feature
- [ ] fix: A bug fix
- [ ] perf: A code change that improves performance
- [ ] refactor: A code change that doesn't fix a bug, or add a feature, or improves performance
- [ ] style: Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)
- [ ] test: Adding missing tests or correcting existing tests

## 3. Deployment tag versioning

Has the version in `params/version.go` been updated?

- [ ] This PR doesn't involve a new deployment, git tag, docker image tag, and it doesn't affect traces
- [ ] Yes

## 4. Breaking change label

Does this PR have the `breaking-change` label?

- [ ] This PR is not a breaking change
- [ ] Yes

### CODE REVIEW

**Review by colinlyguo — APPROVED** [review]

File: rollup/circuitcapacitychecker/impl.go [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+		C.free_c_chars(rawResult)
+	}()
+	log.Debug("ccc set_light_mode end", "id", ccc.ID)
+
+	result := &WrappedCommonResult{}
+	if err := json.Unmarshal([]byte(C.GoString(rawResult)), result); err != nil {
+		return fmt.Errorf("fail to json unmarshal set_light_mode result, id: %d, err: %w", ccc.ID, err)
+	}
+	if result.Error != "" {
+		return fmt.Errorf("fail to set_light_mode in CircuitCapacityChecker, id: %d, err: %w", ccc.ID, result.Error)
```
Comment (mask-pp) [inline rollup/circuitcapacitychecker/impl.go:192]:
```suggestion
		return fmt.Errorf("fail to set_light_mode in CircuitCapacityChecker, id: %d, err: %s", ccc.ID, result.Error)
```

File: rollup/circuitcapacitychecker/impl.go [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+			"len(traces.ExecutionResults)", len(traces.ExecutionResults),
+			"len(traces.TxStorageTraces)", len(traces.TxStorageTraces),
+			"err", "length of Transactions, or ExecutionResults, or TxStorageTraces, is not equal to 1")
+		return nil, ErrUnknown
+	}
+
+	tracesByt, err := json.Marshal(traces)
+	if err != nil {
+		log.Error("fail to json marshal traces in ApplyTransaction", "id", ccc.ID, "TxHash", traces.Transactions[0].TxHash, "err", err)
+		return nil, ErrUnknown
```
Comment (mask-pp) [inline rollup/circuitcapacitychecker/impl.go:71]: Why not return err?

File: rollup/circuitcapacitychecker/impl.go [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+	rawResult := C.apply_tx(C.uint64_t(ccc.ID), tracesStr)
+	defer func() {
+		C.free_c_chars(rawResult)
+	}()
+	log.Debug("check circuit capacity for tx done", "id", ccc.ID, "TxHash", traces.Transactions[0].TxHash)
+
+	result := &WrappedRowUsage{}
+	if err = json.Unmarshal([]byte(C.GoString(rawResult)), result); err != nil {
+		log.Error("fail to json unmarshal apply_tx result", "id", ccc.ID, "TxHash", traces.Transactions[0].TxHash, "err", err)
+		return nil, ErrUnknown
```
Comment (mask-pp) [inline rollup/circuitcapacitychecker/impl.go:89]: ditto

File: rollup/circuitcapacitychecker/impl.go [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+	}
+
+	if result.Error != "" {
+		log.Error("fail to apply_tx in CircuitCapacityChecker", "id", ccc.ID, "TxHash", traces.Transactions[0].TxHash, "err", result.Error)
+		return nil, ErrUnknown
```
Comment (mask-pp) [inline rollup/circuitcapacitychecker/impl.go:94]: ditto

File: rollup/circuitcapacitychecker/impl.go [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+	defer creationMu.Unlock()
+
+	id := C.new_circuit_capacity_checker()
+	ccc := &CircuitCapacityChecker{ID: uint64(id)}
+	ccc.SetLightMode(lightMode)
+	return ccc
+}
+
+// Reset resets a CircuitCapacityChecker
+func (ccc *CircuitCapacityChecker) Reset() {
```
Comment (mask-pp) [inline rollup/circuitcapacitychecker/impl.go:47]:
```suggestion
func Reset(ccc *CircuitCapacityChecker) {
```

File: rollup/circuitcapacitychecker/impl.go [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+	ccc.Lock()
```
Comment (mask-pp) [inline rollup/circuitcapacitychecker/impl.go:48]: Use `creationMu.Lock()` because of `NewCircuitCapacityChecker`.

**Pull request merged**